core/src/typecheck/mod.rs

```diff
@@ -1946,53 +1961,74 @@ fn check<V: TypecheckVisitor>(
             ty.unify(mk_uniftype::dynamic(), state, &ctxt)
                 .map_err(|err| err.into_typecheck_err(state, rt.pos))
         }
-        Term::Let(x, re, rt, attrs) => {
-            let ty_let = binding_type(state, re.as_ref(), &ctxt, true);
+        Term::Let(bindings, rt, attrs) => {
+            // For a recursive let block, shadow all the names we're about to bind, so
+            // we aren't influenced by variables defined in an outer scope.
+            if attrs.rec {
+                for (x, _re) in bindings {
+                    ctxt.type_env
+                        .insert(x.ident(), state.table.fresh_type_uvar(ctxt.var_level));
+                }
+            }
 
-            // We don't support recursive binding when checking for contract equality. See the
-            // `Let` case in `walk`.
-            ctxt.term_env
-                .0
-                .insert(x.ident(), (re.clone(), ctxt.term_env.clone()));
+            let mut tys = Vec::new();
+            let start_ctxt = ctxt.clone();
+            for (x, re) in bindings {
+                let ty_let = binding_type(state, re.as_ref(), &start_ctxt, true);
+
+                // We don't support recursive binding when checking for contract equality. See the
+                // `Let` case in `walk`.
+                ctxt.term_env
+                    .0
+                    .insert(x.ident(), (re.clone(), ctxt.term_env.clone()));
 
-            if attrs.rec {
                 ctxt.type_env.insert(x.ident(), ty_let.clone());
+                visitor.visit_ident(x, ty_let.clone());
+                tys.push((re, ty_let));
             }
 
-            visitor.visit_ident(x, ty_let.clone());
-            check(state, ctxt.clone(), visitor, re, ty_let.clone())?;
-
-            if !attrs.rec {
-                ctxt.type_env.insert(x.ident(), ty_let);
+            let re_ctxt = if attrs.rec { &ctxt } else { &start_ctxt };
+            for (re, ty_let) in tys {
+                check(state, re_ctxt.clone(), visitor, re, ty_let)?;
             }
-            check(state, ctxt, visitor, rt, ty)
+
+            // FIXME: if we're recursive, do we need to do unify the fresh
+            // type variables with the (modified by the recursive check)
+            // binding_type? I feel like we should, but it doesn't seem to make
+            // a difference.
```

Comment:
I'm a bit unsure about this point. I tried unifying the fresh utys, like what `RecRecord` does, but I didn't find any cases where it made a difference.

Reply:
Rereading what `RecRecord` does, it feels a bit over-complicated. I think the approach of this PR is fine, because you store the free unification variable (or type, when there's an annotation) assigned to each binding in `tys`, and use those directly for each `check` later.

The `RecRecord` case does things a bit differently. It enriches the environment in the same way, either with a fresh unification variable or a type for fields with an apparent type. But then, it proceeds to build a record type `{id1: ?a1, ..., idn: ?an}` where the unification variables are entirely free and have no link whatsoever with the one inserted in the environment previously (the difference with your approach). It unifies this record type with the expected type of the whole expression, to first ensure that we are checking a record literal against a record type.

Then, in a second step, it picks the fields that didn't have any annotation/apparent type, and unifies the free-standing variables from the generated record type and what's in the environment: this step is required, because otherwise something like `{foo = 1 +1, bar = foo ++ "a"} : _` would typecheck: `foo` will be bound to `String` in the environment, and the inferred type would be `{foo : Number, bar : String}`, but we'd be missing the constraint that the type `foo : Number` must agree with the inferred binding `foo : String` in the environment.

The only reason I can see why we do that, beside the stupid possibility of overlooking a simpler solution, is as a form of optimization. Indeed, in the recursive record case, when fields are annotated - you can expect that to happen quite often - the current approach eschews some useless identity unifications. Typically, in this PR, with `let rec foo : SuperComplexType = ..., bar : OtherComplexType = ... in body`, we will put `(foo,SuperComplexType)` and `(bar, SuperComplexType)` in `tys`. Then we will check each let-bound variable against this type, which means checking `foo : T` against `T` (after desugaring, `let foo : T = bound, ...` is actually `let foo = (bound : T)`) . This will first unify `T` with itself, which maybe surprisingly isn't free, in particular if `T` is polymorphic. Then it will check `foo` against `T`, as expected.

So, the recursive record approach eschews the additional unification for annotated fields. Even then, honestly, I think we can do it in a much simpler way in both cases: in `tys` (or the hashmap that we are building in the recursive record case), we can use fresh unification variables that are entirely free for fields/bindings that are annotated (because the annotation will take care of the typechecking, and the annotation is also already taken for the recursive environment), while we would use the unification variable inserted in the environment for fields/bindings that aren't annotated, instead of first re-generating entirely free variables for every fields/binding and then in a second step unify this variable with the one in the environment.